### PR TITLE
mavnative: A set of independent changes to improve performance

### PR DESCRIPTION
**Pull request by geeksville**
These changes include independent commits to either improve performance (by supporting --native and --nolog and a few refactorings in mavproxy_link) or to make it easier to profile MAVProxy performance.

### CODE REVIEW

**Comment by geeksville** [comment]
Some numbers from Edison:

# Edison baseline
run 'old' non-optimized mavproxy
* 22% cpu usage - mavproxy.py --master=/dev/ttyMFD1,921600 --rtscts --daemon
* 9% cpu usage - add --streamrate=1 
* 9% cpu usage - add --streamrate=-1 same CPU usage (almost entirely due to idle processing overhead)

# Edison improved
new native version and other optimizations
* 10% with --master=/dev/ttyMFD1,921600 --daemon --native (vs 22% with old version)
* 5% with --master=/dev/ttyMFD1,921600 --daemon --native --nolog
* 4% by removing unused modules
* 2% by using --streamrate=1

**Comment by tridge** [comment]
looks like this needs mavnative, so needs to wait for that to go in

File: MAVProxy/mavproxy.py [hunk 1]
```diff
@@ -280,6 +304,8 @@ def unload_module(modname):
             if hasattr(m, 'unload'):
                 m.unload()
             mpstate.modules.remove((m,pm))
+            mpstate.modules_idle.remove((m,pm))
```
Comment (tridge) [inline MAVProxy/mavproxy.py]: won't this raise ValueError if not in the list?

File: MAVProxy/mavproxy.py [hunk 2]
```diff
@@ -697,7 +738,7 @@ def main_loop():
         for m in mpstate.mav_outputs:
             rin.append(m.fd)
         if rin == []:
-            time.sleep(0.0001)
+            time.sleep(0.1) # We have no ports, check back again in 100ms
```
Comment (tridge) [inline MAVProxy/mavproxy.py]: we'll need to cope with joystick here which needs rapid idle calls

File: MAVProxy/mavproxy.py (hunk 1 above)
Comment (mrpollo) [inline MAVProxy/mavproxy.py]: would a silent failure be best in here, or should we explicitly check for existance before calling remove?

File: MAVProxy/mavproxy.py (hunk 2 above)
Comment (mrpollo) [inline MAVProxy/mavproxy.py]: is there a way to identify joystick so we can add a conditional or any other type of check?

**Comment by mrpollo** [comment]
@tcr3dr I think we should do something about this PR, maybe re-submit a rebased version addressing @tridge concerns?, what do you think?

**Comment by tcr3dr** [comment]
@mrpollo I can definitely handle the improvements to polling / event loops. Let me catch up on all the changes made in this PR over the next day or two and I'll update with a plan.

**Comment by amilcarlucas** [comment]
Will this be picked up, or should it be closed ?

**Comment by khancyr** [comment]
it is really old. 
It looks like have some part here https://github.com/ArduPilot/MAVProxy/pull/556

**Comment by tridge** [comment]
closing as too much has changed, sorry!

**Pull request closed without merging**